### PR TITLE
Fix vars plugin code and documentation mismatch

### LINKED ISSUE #78538: Custom Vars Plugin doesn't work
### Summary

My custom vars plugin is not loaded unless REQUIRES_WHITELIST = False AND the plugin is added to the ANSIBLE_VARS_ENABLED list

### Issue Type

Bug Report

### Component Name

ansible/plugins/loader.py

### Ansible Version

```console
$ ansible --version
ansible [core 2.13.2]
  config file = /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg
  configured module search path = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/lib']
  ansible python module location = /home/noelsc/.local/lib/python3.9/site-packages/ansible
  ansible collection location = /home/noelsc/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/noelsc/.local/bin/ansible
  python version = 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
ANSIBLE_COW_SELECTION(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = random
ANSIBLE_NOCOWS(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = False
CACHE_PLUGIN(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = /home/shared/ansible/cache/
COLOR_VERBOSE(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = bright blue
DEFAULT_ACTION_PLUGIN_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/action_plugins']
DEFAULT_CALLBACK_PLUGIN_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/callback_plugins']
DEFAULT_CONNECTION_PLUGIN_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/connection_plugins']
DEFAULT_FILTER_PLUGIN_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/filter_plugins']
DEFAULT_FORKS(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = 5
DEFAULT_GATHERING(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = implicit
DEFAULT_HOST_LIST(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio/hosts.yml']
DEFAULT_INVENTORY_PLUGIN_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio']
DEFAULT_LOAD_CALLBACK_PLUGINS(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True
DEFAULT_LOCAL_TMP(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = /home/noelsc/.ansible/tmp/ansible-local-7185tquvynhb
DEFAULT_LOG_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = /var/log/ansible.log
DEFAULT_LOOKUP_PLUGIN_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/lookup_plugins']
DEFAULT_MANAGED_STR(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = Ansible managed: {file} modified on %Y-%m-%d %H:%M:%S by {uid} on {host}
DEFAULT_MODULE_PATH(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/lib']
DEFAULT_POLL_INTERVAL(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = 15
DEFAULT_PRIVATE_ROLE_VARS(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True
DEFAULT_TIMEOUT(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = 10
DEFAULT_TRANSPORT(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = smart
DEFAULT_VAULT_PASSWORD_FILE(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = /home/noelsc/.ansible/vault
INVENTORY_ENABLED(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['script', 'host_list', 'yaml', 'constructed', 'community.general.nmap', 'auto']
INVENTORY_IGNORE_EXTS(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ['.pyc', '.pyo', '.swp', '.bak', '~', '.rpm', '.md', '.txt', '~', '.orig', '.ini', '.cfg', '.retry']
INVENTORY_UNPARSED_IS_FAILED(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True
PARAMIKO_HOST_KEY_AUTO_ADD(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True
SHOW_CUSTOM_STATS(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True

CACHE:
=====

jsonfile:
________
_uri(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = /home/shared/ansible/cache

CALLBACK:
========

default:
_______
show_custom_stats(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True

CONNECTION:
==========

paramiko_ssh:
____________
host_key_auto_add(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True

ssh:
___
scp_if_ssh(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = True
timeout(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = 10

SHELL:
=====

sh:
__
remote_tmp(/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg) = ~/.ansible/tmp
```

### OS / Environment

WSL 2.0 Debian 11 and Ubuntu 20.04

### Steps to Reproduce

Here's the custom vars_plugin code.
```yaml (paste below)
from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

DOCUMENTATION = '''
    name: pws_vars
    version_added: "2.13.2"
    short_description: Load vars from PWS
    requirements:
        - Enabled in configuration
    description:
        - Something
'''

import os
from ansible.plugins.vars import BaseVarsPlugin

class VarsModule(BaseVarsPlugin):
    REQUIRES_WHITELIST = False

    def get_vars(self, loader, path, entities, cache=True):
        super(VarsModule, self).get_vars(loader, path, entities)
        self._display.warning('entities {0!r}'.format(entities))
        return {'say_what': 'idk"}
```

Run ansible like this:
```bash
ansible -vvvvvv host.example.com -m debug -a var=level
```

If I change REQUIRES_WHITELIST to True OR leave the custom vars_plugin off the ANSIBLE_VARS_PLUGIN (cli or ansible.cfg file) list, it is not run.

### Expected Results
```
ansible [core 2.13.2]
  config file = /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg
  configured module search path = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/lib']
  ansible python module location = /home/noelsc/.local/lib/python3.9/site-packages/ansible
  ansible collection location = /home/noelsc/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/noelsc/.local/bin/ansible
  python version = 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]
  jinja version = 3.1.2
  libyaml = True
Using /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg as config file
Reading vault password file: /home/noelsc/.ansible/vault
setting up inventory plugins
Loading collection community.general from /home/noelsc/.ansible/collections/ansible_collections/community/general
script declined parsing /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio/hosts.yml as it did not pass its verify_file() method
host_list declined parsing /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio/hosts.yml as it did not pass its verify_file() method
Parsed /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio/hosts.yml inventory source with yaml plugin
Loading collection vivio.passwordstate from /home/noelsc/.ansible/collections/ansible_collections/vivio/passwordstate
Loading callback plugin minimal of type stdout, v2.0 from /home/noelsc/.local/lib/python3.9/site-packages/ansible/plugins/callback/minimal.py
Attempting to use 'default' callback.
Skipping callback 'default', as we already have a stdout callback.
Attempting to use 'junit' callback.
Attempting to use 'minimal' callback.
Skipping callback 'minimal', as we already have a stdout callback.
Attempting to use 'oneline' callback.
Skipping callback 'oneline', as we already have a stdout callback.
Attempting to use 'teams' callback.
Loading callback plugin teams of type notification, v2.0 from /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/callback_plugins/teams.py
Attempting to use 'tree' callback.
[WARNING]: attempts 0
[WARNING]: entities [all]
[WARNING]: entities [audits, host, insnipeit, xenserver]
[WARNING]: entities [host.example.com]
META: ran handlers
host.example.com | SUCCESS => {
    "say_what": "idk"
}
META: ran handlers
META: ran handlers
```
### Actual Results

```console
ansible [core 2.13.2]
  config file = /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg
  configured module search path = ['/mnt/c/Users/Scott.AzureAD/git/vivio-ansible/lib']
  ansible python module location = /home/noelsc/.local/lib/python3.9/site-packages/ansible
  ansible collection location = /home/noelsc/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/noelsc/.local/bin/ansible
  python version = 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]
  jinja version = 3.1.2
  libyaml = True
Using /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/ansible.cfg as config file
Reading vault password file: /home/noelsc/.ansible/vault
setting up inventory plugins
Loading collection community.general from /home/noelsc/.ansible/collections/ansible_collections/community/general
script declined parsing /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio/hosts.yml as it did not pass its verify_file() method
host_list declined parsing /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio/hosts.yml as it did not pass its verify_file() method
Parsed /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/vivio/hosts.yml inventory source with yaml plugin
Loading callback plugin minimal of type stdout, v2.0 from /home/noelsc/.local/lib/python3.9/site-packages/ansible/plugins/callback/minimal.py
Attempting to use 'default' callback.
Skipping callback 'default', as we already have a stdout callback.
Attempting to use 'junit' callback.
Attempting to use 'minimal' callback.
Skipping callback 'minimal', as we already have a stdout callback.
Attempting to use 'oneline' callback.
Skipping callback 'oneline', as we already have a stdout callback.
Attempting to use 'teams' callback.
Loading callback plugin teams of type notification, v2.0 from /mnt/c/Users/Scott.AzureAD/git/vivio-ansible/callback_plugins/teams.py
Attempting to use 'tree' callback.
[WARNING]: attempts 0
META: ran handlers
host.example.com | SUCCESS => {
  "say_what": "VARIABLE IS NOT DEFINED!: 'say_what' is undefined"
}
META: ran handlers
META: ran handlers
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Fixes #78538

Fixes a couple mistakes in the vars plugin developer documentation, as of https://github.com/ansible/ansible/commit/e5ccd18be4d8cec6445f96c2868e0820de0afed8.

* REQUIRES_ENABLED needed to be added to the code and REQUIRES_WHITELIST needed to be deprecated.

* The documentation in `docs/docsite/rst/dev_guide/developing_plugins.rst` was changed to suggest that vars plugins in collections can be run without being enabled (which has never been the case since PluginLoader.all() does not load plugins from collections). I added a warning for it, but of course only occurs if the user actually enabled the plugin... If we don't ever plan on supporting REQUIRES_WHITELIST = False for plugins in collections then the warning is probably overkill and we can just ignore it if it's incorrect.

Marked a WIP until I add a test for the deprecation and make sure there's coverage for legacy plugins.

##### ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request

##### COMPONENT NAME
vars plugins

### CODE REVIEW

File: docs/docsite/rst/dev_guide/developing_collections_structure.rst [hunk 1]
```diff
@@ -94,7 +94,9 @@ plugins directory
 
 Add a 'per plugin type' specific subdirectory here, including ``module_utils`` which is usable not only by modules, but by most plugins by using their FQCN. This is a way to distribute modules, lookups, filters, and so on without having to import a role in every play.
 
-Vars plugins are supported in collections as long as they require being explicitly enabled (using ``REQUIRES_ENABLED``) and they are included using their fully qualified collection name. See :ref:`enable_vars` and :ref:`developing_vars_plugins` for details. Cache plugins may be used in collections for fact caching, but are not supported for inventory plugins.
+Vars plugins are supported in collections. They are not loaded automatically, and always require being explicitly enabled by using their fully qualified collection name. See :ref:`enable_vars` for details.
+
```
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_collections_structure.rst:98]: ~keep a blurb for 'legacy' that requires_enabled?~ nvmd, see it below

File: docs/docsite/rst/dev_guide/developing_collections_structure.rst [hunk 2]
```diff
@@ -94,7 +94,9 @@ plugins directory
 
 Add a 'per plugin type' specific subdirectory here, including ``module_utils`` which is usable not only by modules, but by most plugins by using their FQCN. This is a way to distribute modules, lookups, filters, and so on without having to import a role in every play.
 
-Vars plugins are supported in collections as long as they require being explicitly enabled (using ``REQUIRES_ENABLED``) and they are included using their fully qualified collection name. See :ref:`enable_vars` and :ref:`developing_vars_plugins` for details. Cache plugins may be used in collections for fact caching, but are not supported for inventory plugins.
+Vars plugins are supported in collections. They are not loaded automatically, and always require being explicitly enabled by using their fully qualified collection name. See :ref:`enable_vars` for details.
```
Comment (s-hertel) [inline docs/docsite/rst/dev_guide/developing_collections_structure.rst]:
```suggestion
Vars plugins in collections are not loaded automatically, and always require being explicitly enabled by using their fully qualified collection name. See :ref:`enable_vars` for details.
```

File: lib/ansible/vars/plugins.py [hunk 3]
```diff
@@ -54,8 +56,25 @@ def get_vars_from_path(loader, path, entities, stage):
                 vars_plugin_list.append(vars_plugin)
 
     for plugin in vars_plugin_list:
-        if plugin._load_name not in C.VARIABLE_PLUGINS_ENABLED and getattr(plugin, 'REQUIRES_WHITELIST', False):
-            # 2.x plugins shipped with ansible should require enabling, older or non shipped should load automatically
+        # A collection plugin was enabled to get to this point because vars_loader.all() does not include collection plugins.
+
+        # legacy plugins always run by default, but they can set REQUIRES_ENABLED=True to opt out.
+        # This option does nothing for plugins in collections, but if it's set it should be True to reflect actual behavior.
+        legacy = not '.' in plugin._load_name
```
Comment (s-hertel) [inline lib/ansible/vars/plugins.py]:
```suggestion
        legacy = '.' not in plugin._load_name
```

File: lib/ansible/vars/plugins.py [hunk 4]
```diff
@@ -54,8 +56,25 @@ def get_vars_from_path(loader, path, entities, stage):
...
-            # 2.x plugins shipped with ansible should require enabling, older or non shipped should load automatically
+        # A collection plugin was enabled to get to this point because vars_loader.all() does not include collection plugins.
+
+        # legacy plugins always run by default, but they can set REQUIRES_ENABLED=True to opt out.
+        # This option does nothing for plugins in collections, but if it's set it should be True to reflect actual behavior.
+        legacy = '.' not in plugin._load_name
+
+        # 2.x plugins shipped with ansible should require enabling (host_group_vars is enabled by default for backwards compat).
+        # ansible.legacy should load automatically and run accoring to REQUIRES_ENABLED.
+        needs_enabled = not legacy
```
Comment (s-hertel) [inline lib/ansible/vars/plugins.py]:
```suggestion
        needs_enabled = not legacy
```

**Review by bcoca — APPROVED** [review]

File: lib/ansible/vars/plugins.py [hunk 5]
```diff
@@ -54,8 +56,22 @@ def get_vars_from_path(loader, path, entities, stage):
...
+
+        # legacy plugins always run by default, but they can set REQUIRES_ENABLED=True to opt out.
+        # This option does nothing for plugins in collections, but if it's set it should be True to reflect actual behavior.
+        legacy = '.' not in plugin._load_name
+        needs_enabled = not legacy
+        if hasattr(plugin, 'REQUIRES_WHITELIST'):
+            display.deprecated("The VarsModule class variable 'REQUIRES_WHITELIST' is deprecated. "
+                               "Use 'REQUIRES_ENABLED' instead.", version=2.18)
+            needs_enabled = plugin.REQUIRES_WHITELIST
+        elif hasattr(plugin, 'REQUIRES_ENABLED'):
```
Comment (bcoca) [inline lib/ansible/vars/plugins.py]: make it an if so 'REQUIRES_ENABLED' has precedence

File: lib/ansible/vars/plugins.py [hunk 6]
```diff
@@ -54,8 +56,22 @@ def get_vars_from_path(loader, path, entities, stage):
...
+            needs_enabled = plugin.REQUIRES_ENABLED
+
+        if not legacy and not needs_enabled:
+            display.warning("Vars plugins in collections must be enabled to be loaded, REQUIRES_ENABLED = False is not supported.")
```
Comment (bcoca) [inline lib/ansible/vars/plugins.py]: dont check for needs_enabled, but if either existed, add a 'has_enabled = false' on 64 and set to true  if either property exists, so even when 'false' they should still warn

File: lib/ansible/vars/plugins.py [hunk 7]
```diff
@@ -54,8 +56,22 @@ def get_vars_from_path(loader, path, entities, stage):
...
+        if legacy and plugin._load_name not in C.VARIABLE_PLUGINS_ENABLED and needs_enabled:
```
Comment (bcoca) [inline lib/ansible/vars/plugins.py]: this might need to be an intersection (plugin._load_name, plugin.redirected_names).intersect(ENABLED).length > 0

Comment (bcoca) [reply lib/ansible/vars/plugins.py]: ^ tempted to add to AnsiblePlugin a method  `name_match(listofnames)`

File: lib/ansible/vars/plugins.py (hunk 5 above)
Comment (s-hertel) [reply lib/ansible/vars/plugins.py]: Oh, good idea.

File: lib/ansible/vars/plugins.py (hunk 7 above)
Comment (s-hertel) [reply lib/ansible/vars/plugins.py]: Yes, there's a new FIXME above all() that indicates using _load_name is currently a bug. Since it's pre-existing, I'm going to fix in a folllow up.

File: lib/ansible/vars/plugins.py (hunk 6 above)
Comment (s-hertel) [reply lib/ansible/vars/plugins.py]: I'm not sure what you mean.This only intends to warns if they explicitly set it to False. The implicit default is has_enabled = True.

File: lib/ansible/vars/plugins.py (hunk 5 above)
Comment (s-hertel) [reply lib/ansible/vars/plugins.py]: Fixed.

File: lib/ansible/vars/plugins.py (hunk 6 above)
Comment (s-hertel) [reply lib/ansible/vars/plugins.py]: Okay, added that. I was trying to warn minimally for plugins in collections that were following the dev documentation. I also fixed the existing _load_name bug while I was at it until that's handled in a generic way.

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 8]
```diff
@@ -501,7 +501,9 @@ This ``get_vars`` method just needs to return a dictionary structure with the va
 
 Since Ansible version 2.4, vars plugins only execute as needed when preparing to execute a task. This avoids the costly 'always execute' behavior that occurred during inventory construction in older versions of Ansible. Since Ansible version 2.10, vars plugin execution can be toggled by the user to run when preparing to execute a task or after importing an inventory source.
 
-You can create vars plugins that are not enabled by default using the class variable ``REQUIRES_ENABLED``. If your vars plugin resides in a collection, it cannot be enabled by default. You must use ``REQUIRES_ENABLED`` in all collections-based vars plugins. To require users to enable your plugin, set the class variable ``REQUIRES_ENABLED``:
+Vars plugins in collections are required to be enabled by the user.
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
The user must enable vars plugins in collections .
```

**Review by samccann — COMMENTED** [review]
one little docs nit

File: docs/docsite/rst/dev_guide/developing_plugins.rst (hunk 8 above)
Comment (bcoca) [reply docs/docsite/rst/dev_guide/developing_plugins.rst]: I would rephrase and link the config:
```
- The user must explicitly enable vars plugins that reside in a collection. See :ref:`enabled_vars_plugins` for details.
```

Comment (s-hertel) [reply docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
The user must explicitly enable vars plugins that reside in a collection :ref:`enable_vars`.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 9]
```diff
@@ -501,7 +501,9 @@ This ``get_vars`` method just needs to return a dictionary structure with the va
 
 Since Ansible version 2.4, vars plugins only execute as needed when preparing to execute a task. This avoids the costly 'always execute' behavior that occurred during inventory construction in older versions of Ansible. Since Ansible version 2.10, vars plugin execution can be toggled by the user to run when preparing to execute a task or after importing an inventory source.
 
-You can create vars plugins that are not enabled by default using the class variable ``REQUIRES_ENABLED``. If your vars plugin resides in a collection, it cannot be enabled by default. You must use ``REQUIRES_ENABLED`` in all collections-based vars plugins. To require users to enable your plugin, set the class variable ``REQUIRES_ENABLED``:
+The user must explicitly enable vars plugins that reside in a collection :ref:`enable_vars`.
```
Comment (s-hertel) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
The user must explicitly enable vars plugins that reside in a collection. See :ref:`enable_vars` for details.
```

**Review by samccann — APPROVED** [review]
docs portion LGTM
